pkg/controller/atlasdatafederation/private_endpoint.go

```diff
@@ -1,64 +1,103 @@
 	projectID := project.ID()
-	specPEs := make([]*datafederation.DatafederationPrivateEndpointEntry, 0, len(dataFederation.Spec.PrivateEndpoints))
-	for _, pe := range dataFederation.Spec.PrivateEndpoints {
-		specPEs = append(specPEs, datafederation.NewDatafederationPrivateEndpointEntry(&pe, projectID))
+	fromAtlas, err := service.List(ctx.Context, projectID)
+	if err != nil {
+		return r.privateEndpointsFailed(ctx, err)
 	}
 
-	//NewDatafederationPrivateEndpointEntry
-	atlasPEs, err := getAllDataFederationPEs(ctx.Context, service, projectID)
+	m, err := datafederation.MapDatafederationPrivateEndpoints(projectID, dataFederation, fromAtlas)
 	if err != nil {
-		ctx.Log.Debugw("getAllDataFederationPEs error", "err", err.Error())
+		return r.privateEndpointsFailed(ctx, err)
+	}
+
+	for _, endpoint := range m {
+		endpointReconciler := &PrivateEndpointReconciler{service, endpoint}
+		if err := endpointReconciler.Reconcile(ctx); err != nil {
+			return r.privateEndpointsFailed(ctx, err)
+		}
 	}
 
-	result := syncPrivateEndpointsWithAtlas(ctx, service, projectID, specPEs, atlasPEs)
-	if !result.IsOk() {
-		ctx.SetConditionFromResult(api.DataFederationPEReadyType, result)
-		return result
+	if len(dataFederation.Spec.PrivateEndpoints) == 0 {
+		return r.privateEndpointsUnmanage(ctx)
 	}
 
+	return r.privateEndpointsIdle(ctx)
+}
+
+func (r *AtlasDataFederationReconciler) privateEndpointsFailed(ctx *workflow.Context, err error) workflow.Result {
+	ctx.Log.Errorw("getAllDataFederationPEs error", "err", err.Error())
+	result := workflow.Terminate(workflow.Internal, err.Error())
+	ctx.SetConditionFromResult(api.DataFederationPEReadyType, result)
+	return result
+}
+
+func (r *AtlasDataFederationReconciler) privateEndpointsIdle(ctx *workflow.Context) workflow.Result {
+	ctx.SetConditionTrue(api.DataFederationPEReadyType)
 	return workflow.OK()
 }
 
-func syncPrivateEndpointsWithAtlas(ctx *workflow.Context, service datafederation.DatafederationPrivateEndpointService, projectID string, specPEs, atlasPEs []*datafederation.DatafederationPrivateEndpointEntry) workflow.Result {
-	endpointsToCreate := set.Difference(specPEs, atlasPEs)
-	ctx.Log.Debugw("Data Federation PEs to Create", "endpoints", endpointsToCreate)
-	for _, e := range endpointsToCreate {
-		endpoint := e.(*datafederation.DatafederationPrivateEndpointEntry)
-		if err := service.Create(ctx.Context, endpoint); err != nil {
-			return workflow.Terminate(workflow.Internal, err.Error())
-		}
+func (r *AtlasDataFederationReconciler) privateEndpointsUnmanage(ctx *workflow.Context) workflow.Result {
+	ctx.UnsetCondition(api.DataFederationPEReadyType)
+	return workflow.OK()
+}
+
+type PrivateEndpointReconciler struct {
+	service  datafederation.DatafederationPrivateEndpointService
+	endpoint *datafederation.DataFederationPrivateEndpoint
+}
+
+func (r *PrivateEndpointReconciler) Reconcile(ctx *workflow.Context) error {
```

Comment:
It would be good to point to the state machine diagram doc somehow

Reply:
you mean the dot plot? I think this is a very simple state machine, I suggest to address this as a follow-up if you believe it is worth doing.